pkg/metrics/metrics_test.go

```diff
@@ -315,38 +272,209 @@ func TestWriteMetric(t *testing.T) {
+				}
+
+				w.WriteHeader(http.StatusOK)
+				fmt.Fprintln(w, "ok")
+			}))
+
+			t.Cleanup(ts.Close)
+
 			ctx := context.Background()
-			relativePath := fmt.Sprintf("/%d", rand.Uint64()) //nolint:gosec
+			var relativePath string
 			if tc.responseCodeOverride != 0 {
-				relativePath = fmt.Sprintf("%s/%d", relativePath, tc.responseCodeOverride)
+				relativePath = fmt.Sprintf("%d", tc.responseCodeOverride)
 			}
-			tc.client.Config.ServerURL = fmt.Sprintf("%s%s", ts.URL, relativePath)
+			tc.client.Config.ServerURL = fmt.Sprintf("%s/%s", ts.URL, relativePath)
 
 			err := tc.client.WriteMetric(ctx, tc.metric, tc.count)
 
 			if diff := testutil.DiffErrString(err, tc.wantErr); diff != "" {
 				t.Error(diff)
 			}
-			val, ok := reqMap.Load(relativePath + "/sendMetrics")
 			if tc.wantRequest != nil {
-				if !ok {
+				if saveReq == nil {
 					t.Errorf("no http request received, expected body of: %v", *tc.wantRequest)
 				}
-				request, ok := val.(*http.Request)
-				if !ok {
-					t.Fatal("Expected *httpRequest in sync map, but cast failed.")
+				defer saveReq.Body.Close()
+
+				var got SendMetricRequest
+				if err := json.NewDecoder(saveReq.Body).Decode(&got); err != nil {
+					t.Errorf("error reading request to test server: %s", err.Error())
+				}
+				if diff := cmp.Diff(&got, tc.wantRequest); diff != "" {
+					t.Errorf("unexpected request body. Diff (-got +want): %s", diff)
+				}
+			} else {
+				if saveReq != nil {
+					t.Errorf("did not expect a request but got one")
+				}
+			}
+		})
+	}
+}
+
+func TestWriteMetricAsync(t *testing.T) {
+	t.Parallel()
+
+	cases := []struct {
+		name          string
+		metric        string
+		count         int64
+		client        *client
+		timeoutMillis int
+		wantRequest   *SendMetricRequest
+		wantErr       string
+	}{
+		{
+			name:   "metric_success",
+			metric: "foo",
+			count:  1,
+			client: defaultClient(),
+			wantRequest: &SendMetricRequest{
+				AppID:      testAppID,
+				AppVersion: testVersion,
+				Metrics:    map[string]int64{"foo": 1},
+				InstallID:  testInstallID,
+			},
+		},
+		{
+			name:          "metric_success_timeout_set",
+			metric:        "foo",
+			count:         1,
+			client:        defaultClient(),
+			timeoutMillis: 4000,
+			wantRequest: &SendMetricRequest{
+				AppID:      testAppID,
+				AppVersion: testVersion,
+				Metrics:    map[string]int64{"foo": 1},
+				InstallID:  testInstallID,
+			},
+		},
+		{
+			name:   "metric_opt_out_noop",
+			metric: "foo",
+			count:  1,
+			client: func() *client {
+				c := defaultClient()
+				c.OptOut = true
+				return c
+			}(),
+			wantRequest: nil,
+		},
+		{
+			name:          "metric_failure_timeout",
+			metric:        "foo",
+			count:         1,
+			client:        defaultClient(),
+			timeoutMillis: 9,
+			wantRequest:   nil,
+			wantErr:       "context deadline exceeded",
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			var saveReqMutex sync.Mutex
```

Comment:
Is the mutex still necessary?

Reply:
Yes, there is a data race without it (wouldn't happen in practice but golang detects it making ci fail)